core/src/com/unciv/logic/map/TileInfo.kt

```diff
@@ -376,22 +379,35 @@ open class TileInfo {
             improvement.name == "Railroad" && this.roadStatus != RoadStatus.Railroad && !isWater -> true
             improvement.name == "Remove Road" && this.roadStatus == RoadStatus.Road -> true
             improvement.name == "Remove Railroad" && this.roadStatus == RoadStatus.Railroad -> true
-            topTerrain.unbuildable && (topTerrain.name !in improvement.resourceTerrainAllow) -> false
+            topTerrain.unbuildable && !improvement.isAllowedOnFeature(topTerrain.name) -> false
             // DO NOT reverse this &&. isAdjacentToFreshwater() is a lazy which calls a function, and reversing it breaks the tests.
             improvement.hasUnique("Can also be built on tiles adjacent to fresh water") && isAdjacentToFreshwater -> true
             "Can only be built on Coastal tiles" in improvement.uniques && isCoastalTile() -> true
+            improvement.uniqueObjects.filter { it.placeholderText == "Can only be built on [] tiles" }.any {
+                unique -> !matchesUniqueFilter(unique.params[0])
+            } -> false
             else -> resourceIsVisible && getTileResource().improvement == improvement.name
         }
     }
 
+    /**
+     * Implementation of _`tileFilter`_ 
+     * @see <a href="https://github.com/yairm210/Unciv/wiki/uniques#user-content-tilefilter">tileFilter</a>
+     */
     fun matchesUniqueFilter(filter: String, civInfo: CivilizationInfo? = null): Boolean {
-        return filter == baseTerrain
+        return filter == "All"
+                || '+' in filter && filter.split('+').all { matchesUniqueFilter(it, civInfo) }
+                || ',' in filter && filter.split(',').any { matchesUniqueFilter(it, civInfo) }
```

Comment:
Such uniques won't be translated

Reply:
I mentioned that in "Problems to resolve" - That was asking for ideas: teach the translator or find another way to express such filters?